PnP update

marip8 opened the pull request:
Per investigation in #57, this PR provides a workaround for using Eigen objects without local paramterization. It also adds a function for calculating the "full" covariance matrix between two parameter blocks (i.e. variance with the parameter block itself and covariance with the other parameter block)

marip8 commented:
@schornakj can you review? This is ready to merge but has a few changes that might overlap with #62

schornakj commented on rct_optimizations/src/rct_optimizations/pnp.cpp:
```diff
@@ -53,16 +53,15 @@ struct SolvePnPCostFunc3D
     using Vector3 = Eigen::Matrix<T, 3, 1>;
 
-    Eigen::Map<const Eigen::Quaternion<T>> q(target_q);
+    Eigen::Map<const Vector3> aa(target_aa);
```
Nitpicky style preference thing, but I'd prefer if `aa` and `t` were named more verbosely so they can be understood at a glance without needing to follow them through the code.

schornakj commented on rct_optimizations/src/rct_optimizations/pnp.cpp:
```diff
@@ -85,31 +84,24 @@ namespace rct_optimizations
   // Create the optimization variables from the input guess
-  Eigen::Quaterniond q(params.camera_to_target_guess.rotation());
+  Eigen::AngleAxisd rot(params.camera_to_target_guess.rotation());
+  Eigen::Vector3d aa = rot.angle() * rot.axis();
```
Same comment as above.

schornakj commented on rct_optimizations/test/pnp_utest.cpp:
```diff
@@ -126,22 +142,14 @@ TEST(PNP_3D, PerfectInitialConditions)
   EXPECT_LT(result.final_cost_per_obs, 1.0e-15);
+  checkCorrelation(result.camera_to_target_covariance);
+
+//  Eigen::IOFormat fmt(4, 0, " | ", "\n", "|", "|");
+//  std::cout << "Covariance:\n" << result.camera_to_target_covariance.format(fmt) << std::endl;
```
Do you want to leave this commented out? Might be worthwhile to print during the tests anyway.

schornakj commented:
> @schornakj can you review? This is ready to merge but has a few changes that might overlap with #62

I'll rebase onto your changes after we merge. #62 adds a function that's similar (but uses different underlying Ceres functions) to the "full" covariance function you added here so you might be interested in taking a look and seeing which approach you prefer.

schornakj commented on rct_optimizations/test/pnp_utest.cpp:
```diff
@@ -159,7 +159,7 @@ TEST_F(PnP3DTest, PerturbedInitialCondition)
   EXPECT_TRUE(result.converged);
   EXPECT_TRUE(result.camera_to_target.isApprox(target_to_camera.inverse(), 1.0e-6));
-  EXPECT_LT(result.final_cost_per_obs, 1.0e-12);
+  EXPECT_LT(result.final_cost_per_obs, 1.0e-10);
```
I think we need a better way to evaluate these variable costs in our tests. Maybe we should compare costs across multiple optimizations and check that the mean and variance of all the costs are within some specified range.

marip8 commented:
@schornakj I addressed your comments in the latest commits. I ended up running the randomly initialized optimizations 30 times and setting expectations that the mean error + 3 standard deviations should be less than a (still somewhat arbitrary) threshold. I think this is at least a step in the right direction for setting reasonable expectations for unit tests with randomness

schornakj commented:
> @schornakj I addressed your comments in the latest commits. I ended up running the randomly initialized optimizations 30 times and setting expectations that the mean error + 3 standard deviations should be less than a (still somewhat arbitrary) threshold. I think this is at least a step in the right direction for setting reasonable expectations for unit tests with randomness

That's reasonable. Looks good to me!
